tools/topology/platform/intel/intel-generic-dmic.m4

```diff
@@ -0,0 +1,98 @@
+#
+# Topology for digital microphones array
+#
+
+# Include topology builder
+include(`utils.m4')
+include(`dai.m4')
+include(`pipeline.m4')
+
+# Include TLV library
+include(`common/tlv.m4')
+
+# Include Token library
+include(`sof/tokens.m4')
+
+include(`platform/intel/bxt.m4')
+include(`platform/intel/dmic.m4')
+
+# defined in machine driver
+define(DMIC_DAI_LINK_32_ID, `6')
+define(DMIC_DAI_LINK_16_ID, `7')
+
+define(DMIC_PIPELINE_32_ID, `100')
+define(DMIC_PIPELINE_16_ID, `101')
+
+#
+# Define the pipelines
+#
+
+dnl PIPELINE_PCM_ADD(pipeline,
+dnl     pipe id, pcm, max channels, format,
+dnl     frames, deadline, priority, core)
+
+# Passthrough capture pipeline using max channels defined by CHANNELS.
+# Schedule 48 frames per 1000us deadline on core 0 with priority 0
+PIPELINE_PCM_ADD(sof/pipe-passthrough-capture.m4,
+	DMIC_PIPELINE_32_ID, DMIC_DAI_LINK_32_ID, CHANNELS, s32le,
+	48, 1000, 0, 0)
+
+# Passthrough capture pipeline using max channels defined by CHANNELS.
+# Schedule 48 frames per 1000us deadline on core 0 with priority 0
+PIPELINE_PCM_ADD(sof/pipe-passthrough-capture.m4,
+	DMIC_PIPELINE_16_ID, DMIC_DAI_LINK_16_ID, CHANNELS, s16le,
+	48, 1000, 0, 0)
+
+#
+# DAIs configuration
+#
+
+dnl DAI_ADD(pipeline,
+dnl     pipe id, dai type, dai_index, dai_be,
+dnl     buffer, periods, format,
+dnl     frames, deadline, priority, core)
+
+# capture DAI is DMIC 0 using 2 periods
+# Buffers use s32le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+	DMIC_PIPELINE_32_ID, DMIC, 0, dmic32,
+	concat(`PIPELINE_SINK_', DMIC_PIPELINE_32_ID), 2, s32le,
+	48, 1000, 0, 0)
+
+# capture DAI is DMIC 1 using 2 periods
+# Buffers use s16le format, with 48 frame per 1000us on core 0 with priority 0
+DAI_ADD(sof/pipe-dai-capture.m4,
+	DMIC_PIPELINE_16_ID, DMIC, 1, dmic16,
+	concat(`PIPELINE_SINK_', DMIC_PIPELINE_16_ID), 2, s16le,
+	48, 1000, 0, 0)
+
+
+dnl PCM_DUPLEX_ADD(name, pcm_id, playback, capture)
+dnl PCM_CAPTURE_ADD(name, pipeline, capture)
+PCM_CAPTURE_ADD(DMIC32, DMIC_DAI_LINK_32_ID, concat(`PIPELINE_PCM_', DMIC_PIPELINE_32_ID))
+PCM_CAPTURE_ADD(DMIC16, DMIC_DAI_LINK_16_ID, concat(`PIPELINE_PCM_', DMIC_PIPELINE_16_ID))
+
+#
+# BE configurations - overrides config in ACPI if present
+#
+
+dnl DAI_CONFIG(type, dai_index, link_id, name, ssp_config/dmic_config)
+ifelse(CHANNELS, 4,
```

Comment:
Please comment this conditional logic. Where is CHANNELS expected to be defined ? Please comment.

Reply:
FW could be configured by make menuconfig. I also want to set CHANNELS in kconfig. Now it is not ready for topology, but Poland team is doing a task to dynamically configure topology. what is your idea ?

Reply:
another solution is to define it in CMakeList.txt like, change it to 2 or 4 when you need
````
"sof-hda-generic\;sof-hda-generic;-DCHANNELS=2"
````

Reply:
I think the problem is that you are trying to use one topology file for two **different** machines. The solution is to create **two** topology files, one for each machine that will define CHANNELS, PCM_IDs etc and include the hda.m4 and dmic.m4

Reply:
Thanks! I will update it later